flexible-event/flexible_event_privacy.py

```diff
@@ -0,0 +1,167 @@
+import argparse
+import collections
+import functools
+import json
+from math import log, exp
+import sys
+from typing import Tuple, TypedDict, List, NamedTuple
+
+
+PerTriggerDataConfig = List[Tuple[int, int]]
+class ApiConfig(NamedTuple):
+    max_event_level_reports: int
+    per_trigger_data_configs: PerTriggerDataConfig
+
+
+def num_flexible_states(config: ApiConfig) -> int:
+    """Returns the total number of output states for a given configuration of
+       the flexible event-level API
+
+    Args:
+    max_event_level_reports: The value of `max_event_level_reports` in the source registration.
+    per_type_configs: A list of tuples of (num_windows, num_summary_buckets), per `trigger_data`
+    """
+
+    # Let B be the trigger data cardinality.
+    # For every trigger data i, there are w_i windows and c_i maximum reports / summary buckets.
+    # The following helper function memoizes the recurrence relation:
+    # 1. A[C, w_1, ..., w_B, c_1, ... , c_B] = 1 if B = 0
+    # 2. A[C, w_1, ..., w_B, c_1, ... , c_B] = A[C, w_1, ..., w_{B-1}, c_1, ... , c_{B-1}] if w_B = 0
+    # 3. A[C, w_1, ..., w_B, c_1, ... , c_B] = sum(A[C - j, w_1, ..., w_B - 1, c_1, ... , c_B - j], j from 0 to min(c_B, C)) otherwise
+    @functools.lru_cache(maxsize=None)
+    def helper(total_cap: int, index: int, w: int, c: int) -> int:
+        # Case 1.
+        if index == 0 and w == 0:
+            return 1
+
+        # Case 2.
+        if w == 0:
+            trigger_config = config.per_trigger_data_configs[index - 1]
+            return helper(total_cap, index - 1, trigger_config[0], trigger_config[1])
+
+        # Case 3.
+        return sum(helper(total_cap - i, index, w - 1, c - i) for i in range(min(c, total_cap) + 1))
+
+    last_config = config.per_trigger_data_configs[-1]
+    data_cardinality = len(config.per_trigger_data_configs)
+    return helper(config.max_event_level_reports, data_cardinality - 1, last_config[0], last_config[1])
+
+
+def h(x: float) -> float:
+    """Evaluates the binary entropy function.
+
+    Args:
+    x: the input value.
+
+    Returns:
+      The binary entropy function at x.
+    """
+    if x == 0 or x == 1:
+        return 0
+    else:
+        return - x * log(x, 2) - (1 - x) * log(1 - x, 2)
+
+
+def flip_probability_dp(num_states: int, epsilon: float) -> float:
```

Comment:
(TIL epsilon could be a float)

Reply:
Indeed, it has meaning as long as its a real number >= 0 (epsilon = 0 means you don't leak any information at all).